src/main/java/com/mongodb/jdbc/MongoJsonSchema.java

```diff
@@ -4,22 +4,216 @@
+                    // Move the items down with the anyOf schema for the bsontype Array
+                    // because they go together
+                    anyOfSchema.items = toSimplifiedMongoJsonSchema(baseSchema.items);
+                } else if (BSON_OBJECT.getBsonName().equalsIgnoreCase(currType)) {
+                    // Move the object related properties down with the anyof schema for the 'object'
+                    anyOfSchema.properties = toMongoJsonSchemaProperties(baseSchema.properties);
+                    anyOfSchema.required = baseSchema.required;
+                    anyOfSchema.additionalProperties = baseSchema.additionalProperties;
+                    result.properties = null;
+                    result.required = null;
+                    result.additionalProperties = false;
+                }
+                // Add the bsontype information as a new anyOf schema
+                if (!result.anyOf.contains(anyOfSchema)) {
+                    result.anyOf.add(anyOfSchema);
+                }
+            }
+        }
+
+        return result;
+    }
+
+    /**
+     * Converts a polymorphic bsonType which can either be a BsonArray or a BsonString to a set of
+     * Strings.
+     *
+     * @param polymorphicBsonType The original polymorphic type.
+     * @return the corresponding String set.
+     */
+    private static Set<String> polymorphicBsonTypeToStringSet(BsonValue polymorphicBsonType) {
+        Set<String> result;
+        if (polymorphicBsonType.isArray()) {
+            result =
+                    polymorphicBsonType
+                            .asArray()
+                            .stream()
+                            .map(val -> val.asString().getValue())
+                            .collect(Collectors.toSet());
+        } else if (polymorphicBsonType.isString()) {
+            result = new HashSet<String>();
+            result.add(polymorphicBsonType.asString().getValue());
+        } else {
+            throw new BsonInvalidOperationException(
+                    "Value expected to be of type "
+                            + BsonType.ARRAY
+                            + " or "
+                            + BsonType.STRING
+                            + " but  is of unexpected type "
+                            + polymorphicBsonType.getBsonType());
+        }
+
+        return result;
+    }
+
+    /**
+     * Flattens nested anyOf.
+     *
+     * @param ioSchema The schema to simplify. ioSchema will be modified directly.
+     * @return the simplified schema for convenience.
+     */
+    private static MongoJsonSchema flattenNestedAnyOfs(MongoJsonSchema ioSchema) {
+        if (null == ioSchema) {
+            return null;
+        }
+
+        MongoJsonSchema result = ioSchema;
+        if (result.anyOf != null && !result.anyOf.isEmpty()) {
+            result.anyOf =
+                    result.anyOf
+                            .stream()
+                            .flatMap(
```

Comment:
<3 `flatMap`. Combinators just make code so much easier to follow once you learn them.